// @traceo-sdk/node/dist/metrics/default/heap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeapMetrics = void 0;
const node_core_1 = require("@traceo-sdk/node-core");
/**
 * https://www.geeksforgeeks.org/node-js-v8-getheapstatistics-method/
 *
 */
const TO_MB = 1024 * 1024;
class HeapMetrics {
    constructor() { }
    collect() {
        return [
            {
                name: "heap_used",
                type: node_core_1.TraceoMetricType.GAUGE,
                unixTimestamp: node_core_1.utils.currentUnix(),
                value: this.usedHeap
            },
            {
                name: "heap_total",
                type: node_core_1.TraceoMetricType.GAUGE,
                unixTimestamp: node_core_1.utils.currentUnix(),
                value: this.totalHeap
            },
            {
                name: "heap_rss",
                type: node_core_1.TraceoMetricType.GAUGE,
                unixTimestamp: node_core_1.utils.currentUnix(),
                value: this.rss
            }
        ];
    }
    get usedHeap() {
        return node_core_1.utils.toDecimalNumber(process.memoryUsage().heapUsed / TO_MB);
    }
    get totalHeap() {
        return node_core_1.utils.toDecimalNumber(process.memoryUsage().heapTotal / TO_MB);
    }
    get rss() {
        return node_core_1.utils.toDecimalNumber(process.memoryUsage().rss / TO_MB);
    }
}
exports.HeapMetrics = HeapMetrics;
